docs/grammar/Solidity.g4

```diff
@@ -0,0 +1,489 @@
+importAliases: symbol=identifier (As alias=identifier)?;
+/**
+ * Path of a file to be imported.
+ */
+path: NonEmptyStringLiteral;
+/**
+ * List of aliases for symbols to be imported.
+ */
+symbolAliases: LBrace aliases+=importAliases (Comma aliases+=importAliases)* RBrace;
+
+/**
+ * Top-level definition of a contract.
+ */
+contractDefinition:
+	Abstract? Contract name=identifier
+	inhertianceSpecifierList?
+	LBrace contractBodyElement* RBrace;
+/**
+ * Top-level definition of an interface.
+ */
+interfaceDefinition:
+	Interface name=identifier
+	inhertianceSpecifierList?
+	LBrace contractBodyElement* RBrace;
+/**
+ * Top-level definition of a library.
+ */
+libraryDefinition: Library name=identifier LBrace contractBodyElement* RBrace;
+
+//@doc:inline
+inhertianceSpecifierList:
+    Is inheritanceSpecifiers+=inheritanceSpecifier
+    (Comma inheritanceSpecifiers+=inheritanceSpecifier)*?;
+/**
+ * Inheritance specifier for contracts and interfaces.
+ * Can optionally supply base constructor arguments.
+ */
+inheritanceSpecifier: name=userDefinedTypeName arguments=callArgumentList?;
+
+/**
+ * Declarations that can be used in contracts, interfaces and libraries.
+ *
+ * Note that interfaces and libraries may not contain state variables and constructors and libraries additionally
+ * may not contain fallback or receive functions.
+ */
+contractBodyElement:
+    constructorDefinition
+    | functionDefinition
+    | modifierDefinition
+    | fallbackReceiveFunctionDefinition
+    | structDefinition
+    | enumDefinition
+    | stateVariableDeclaration
+    | eventDefinition
+    | usingDirective;
+//@doc:inline
+namedArgument: name=identifier Colon value=expression;
+/**
+ * Arguments when calling a function or a similar callable object.
+ * The arguments are either given as comma separated list or as map of named arguments.
+ */
+callArgumentList: LParen ((expression (Comma expression)*)? | LBrace (namedArgument (Comma namedArgument)*)? RBrace) RParen;
+/**
+ * Qualified name of a user defined type.
+ */
+userDefinedTypeName: identifier (Period identifier)*;
+
+/**
+ * Call to a modifier. If the modifier takes no arguments, the argument list can be skipped entirely
+ * (including opening and closing parentheses).
+ */
+modifierInvocation: identifier callArgumentList?;
+/**
+ * Visibility for functions and function types.
+ */
+visibility: Internal | External | Private | Public;
+/**
+ * A list of parameters, such as function arguments or return values.
+ */
+parameterList: parameters+=parameterDeclaration (Comma parameters+=parameterDeclaration)*;
+//@doc:inline
+parameterDeclaration: type=typeName location=storageLocation? name=identifier?;
+/**
+ * Definition of a constructor.
+ * Must always supply an implementation.
+ * Note that specifying internal or public visibility is deprecated.
+ */
+constructorDefinition:
+    Constructor LParen (arguments=parameterList)? RParen
+    (modifierInvocation | Payable | Internal | Public)*
```

Comment:
Having this as `(...)*` implies that one could repeat them, doesn't it?

Reply:
Yep. The reason for doing that is that the order isn't fixed.
Interestingly, we apparenlty don't have tests for e.g. ``constructor() payable payable``, otherwise this would have failed them :-). I'll add the ugly ``visibilitySet`` and ``payableSet`` locals to this as well, then it'll be precise.

Reply:
Also added tests.